Improvement of the TLS protocol

WagnerPMC opened the pull request:
## Reason for this PR
Issue #598

## Description of Changes
1. New setting in the configuration file is `general.trust_os_certificates`. Which means whether the developer can trust not only the built-in certificates in PGCat, but also the OS root certificates
2. Optional `only-ca` mode to configure `general.verify_server_certificate`. It is equivalent to the `sslmode=verify-ca` mode

## License Acceptance
By submitting this pull request, I confirm that my contribution is made under
the terms of the [PGCat license](https://github.com/postgresml/pgcat/blob/main/LICENSE).

WagnerPMC commented:
@levkk, Why is CI crashing on std::sync::OnceLock? This feature was stabilized in Rust 1.70.

WagnerPMC commented:
@levkk , I think I found the cause. The Dockerfile.ci file uses Rust 1.67, so the compilation fails. Is there any reason why CI uses Rust 1.67?

levkk commented:
We use `once_cell` elsewhere in the code, so it's not a bad idea to use it here too. We can move to `OnceLock` at some point later, for every usage of `once_cell`.

WagnerPMC commented:
> We use `once_cell` elsewhere in the code, so it's not a bad idea to use it here too. We can move to `OnceLock` at some point later, for every usage of `once_cell`.

What do you think about my PR?

WagnerPMC commented:
Yeah, I think I will, but in a slightly different way. So that during each hot reload of the server (if configured), the operating system certificate cache is also reset.

WagnerPMC commented:
I'm not going to implement everything in one pull request. I can accept this one to start with, then I can create a second one for the next opportunity I wrote about above. **So I declare the PR to be final.**

levkk commented on CONFIG.md:
````diff
@@ -63,24 +63,53 @@ path: general.server_round_robin
+
+### trust_os_certificates
+```
+path: general.trust_os_certificates
````
I think this should be default true. Also, this makes me think, do we need a setting for this at all? When would one not want to trust the OS certificates? The inclusion of webpki wasn't done intentionally, it was just part of all the examples for rustls. I think this is the desired way to use TLS certificates with pgcat, not with webpki. I think I would remove this option and load OS certificates by default.

WagnerPMC replied:
I was thinking the same thing. If you think so, I share your point of view.

levkk commented on src/config.rs:
```diff
@@ -254,6 +255,36 @@ impl User {
+#[serde(untagged)]
+pub enum CertificateVerificationVariant {
+    Bool(bool),     // false -> prefer, true - verify-full
+    String(String), // "only-ca" -> verify-full
```
"verify-ca" is an actual mode supported by the client, we should call it that way here too: https://www.postgresql.org/docs/current/libpq-ssl.html

WagnerPMC replied:
> "verify-ca" is an actual mode supported by the client, we should call it that way here too: https://www.postgresql.org/docs/current/libpq-ssl.html

I'll correct that oversight

levkk commented on src/server.rs:
```diff
@@ -407,20 +410,57 @@ impl Server {
+                                debug!("{} get_os_root_certificates processed {} valid and {} invalid certs", address, result.0, result.1);
+                            }
+                            Err(err) => {
+                                warn!("{} Failed to load OS root certificates: {:?}", address, err)
```
This would log a lot. I think we should try to load the store on boot, e.g. during config validation, and only issue a warning once.

levkk commented on src/server.rs:
```diff
@@ -407,20 +410,57 @@ impl Server {
+                                    // verify-ca
+                                    dangerous.set_certificate_verifier(Arc::new(
+                                        crate::tls::OnlyRootCertificateVerification {
+                                            roots: root_store.clone(),
```
This is a large structure, according to the docs. Can we pass a reference instead?

WagnerPMC replied:
> This is a large structure, according to the docs. Can we pass a reference instead?

I think yes, it can be converted to a reference

WagnerPMC commented:
@levkk ,

Please review it now.
A nice bonus: by using OS certificates, the binary file size is reduced by about 750KB.

For some reason CI crashes with an error:
```
error: package `clap_lex v0.5.1` cannot be built because it requires rustc 1.70.0 or newer, while the currently active rustc version is 1.67.1
```

@levkk, please check the dependencies

levkk commented on src/tls.rs:
```diff
@@ -7,14 +7,58 @@ use std::sync::Arc;
+/// from anywhere in the app.
+/// ArcSwap makes this cheap and quick.
+pub fn get_root_cert_store() -> RootCertStore {
+    (*(*ROOT_CERT_STORE.load())).clone()
```
A clone is expensive, I think you want to pass by reference here and drop that reference as quickly as possible after validating the certificate.

WagnerPMC replied:
Then please tell me how I can competently take a reference to this object?

WagnerPMC replied:
> Then please tell me how I can competently take a reference to this object?

I wrote that for a reason. You have a similar mechanism used in your entire code

levkk commented on src/main.rs:
```diff
@@ -187,7 +196,23 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         };
 
+        {
+            let mut autoreload_tls_interval = tokio::time::interval(tokio::time::Duration::from_secs(60 * 60)); // update root certificates every hour
```
I'm not sure we want this. Do you have very short-lived certificates in your setup? Typical certificates are issued for at least 3 months (e.g. let's encrypt) and private CAs can issue arbitrary length certs. In any case, this shouldn't be done automatically every second but with a command issued to the pooler instead, e.g. you can do this when we issue a `RELOAD` to the admin database.

WagnerPMC replied:
What makes you think the certificates will be updated every second? I made them update hourly.

levkk replied:
Sorry, that was a typo on my part. Every hour. But even that seems odd, that's not something that happens as far as I know, and even if some environments issue very short-lived certs, I think we should make this an explicit command for Pgcat to reload them, and not have to wait every hour. Some environments for example may choose to rotate certs more frequently, and the hourly frequency is not often enough for them. So my proposal is reload certs when someone issues a `RELOAD` command to the pooler or a SIGHUP, basically in `reload_config`.

WagnerPMC replied:
I agree with you. Good, then I will correct it and prepare a new version

levkk commented on src/main.rs:
```diff
@@ -92,6 +92,15 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
+                Err(_) => {
+                    // no need for the error! macro, because the reload_root_cert_store()
+                    // function already uses the error! macro.
+                    std::process::exit(exitcode::OSFILE);
```
Not entirely sure you want to exit here, some setups may not want TLS or don't have the certs installed at all which could cause an error perhaps? I'm not entirely sure why the root store could give an error, but in any case, if TLS is optional, which it can be if we set `server_tls` to false, this should not stop the pooler from starting.

WagnerPMC replied:
Okay, I'll think about it.

WagnerPMC commented:
@levkk, I did everything you mentioned.

Currently, loading root certificates on PGCat startup will not occur unless the `verify_server_certificate` setting is configured.
Root certificates will only be loaded in the following cases:
1. SIGHUP signal is given to the process control
2. The RELOAD command in the PGCat console was executed.
3. If server certificate validation was disabled before the configuration was automatically rebooted and enabled after the reboot.

I've removed the extra cloning, but we still have cloning happening in a single location: https://github.com/WagnerPMC/pgcat/blob/fc8998f0d16aaf53ec8a64685c10dd55a3b654b2/src/server.rs#L403

And unfortunately, there's no getting away from it. This is because rustls must borrow the RootCertStore. So we have no power to influence it here.

In support of this PR, I will say that constant cloning (in implicit form) has happened before: https://github.com/postgresml/pgcat/blob/a054b454d2566e1a90ef373c7a5360850f12484e/src/server.rs#L399-L412

This concludes my commentary. I have changed the PR header. @levkk, review it now

WagnerPMC commented:
I think PR is ready. 
@levkk , review it please

WagnerPMC commented:
> We need to figure out why this doesn't build in CI and then we can probably merge this as is. Without CI, we haven't tested this PR yet.
> 
> It looks like Cargo.lock updated several of the dependencies. Have you ran `cargo update` by any chance when you added the the new dependency? Ideally, you would just add it to `Cargo.toml` and run `cargo build` ensuring that no existing dependencies are updated.

Yes, you're right. My mistake was that I did a ```cargo update```.
I reverted back to the previous versions of the dependencies, now CI runs successfully.

levkk commented on src/config.rs:
```diff
@@ -1456,10 +1498,10 @@ pub async fn reload_config(client_server_map: ClientServerMap) -> Result<bool, E
 
     if old_config != new_config {
-        info!("Config changed, reloading");
-        ConnectionPool::from_config(client_server_map).await?;
```
Why did you remove this from here and copy/pasted it in 3 other locations instead? It seems like you'd want to just reload the connection pools in every case now, which is fine, since we don't re-initialize connection pools if the configuration hasn't changed. Maybe we should just call `ConnectionPool::from_config(...)` here and not call it independently everywhere else in the code?

WagnerPMC replied:
> Why did you remove this from here and copy/pasted it in 3 other locations instead? It seems like you'd want to just reload the connection pools in every case now, which is fine, since we don't re-initialize connection pools if the configuration hasn't changed. Maybe we should just call `ConnectionPool::from_config(...)` here and not call it independently everywhere else in the code?

Since certificates can be renewed, isn't it required to reload the pools afterwards?

WagnerPMC replied:
@levkk, please reply me

levkk replied:
Apologies, been very busy this week. I'll take another look at this. I think this is pretty much ready to go, just need to give it one more look.

WagnerPMC replied:
> Apologies, been very busy this week. I'll take another look at this. I think this is pretty much ready to go, just need to give it one more look.

Did you watch my PR?

levkk replied:
My question is not why you need to reload pools but why did you remove the code that does it automatically and copy/pasted it in 3 different places that do the same thing as before?

levkk replied:
https://github.com/WagnerPMC/pgcat/pull/1

WagnerPMC commented:
@levkk , Can you tell me why your Pull Request contains falling code?
